### PR TITLE
add withTimeout method to sql"" and SQL("") queries.

### PR DESCRIPTION
**Pull request by gregghz**
It throws and exception and "gracefully" kills the query in the event of a timeout.

### CODE REVIEW

File: src/main/scala/com/lucidchart/open/relate/StatementPreparer.scala [hunk 1]
```diff
@@ -61,6 +61,12 @@ private[relate] sealed trait StatementPreparer {
 private[relate] trait BaseStatementPreparer extends StatementPreparer {
   protected def applyParams(stmt: PreparedStatement)
   protected def parsedQuery: String
+  protected def timeout: Option[Int] = None
+
+  protected def setTimeout(stmt: PreparedStatement): Unit = for {
+    seconds <- timeout
+    stmt <- Option(stmt)
+  } yield (stmt.setQueryTimeout(seconds))
```
Comment (pauldraper) [inline src/main/scala/com/lucidchart/open/relate/StatementPreparer.scala]: Can't we just do `timeout.foreach(stmt.setQueryTimeout)`?

Comment (pauldraper) [inline src/main/scala/com/lucidchart/open/relate/StatementPreparer.scala]: Oh, this is in case `stmt` is `null`. Will that ever happen?

Comment (shinpatsy) [inline src/main/scala/com/lucidchart/open/relate/StatementPreparer.scala]: that's right. in testing, it came up null a few times.

Comment (gregghz) [inline src/main/scala/com/lucidchart/open/relate/StatementPreparer.scala]: It happened when I was testing on the user service.

Comment (shinpatsy) [inline src/main/scala/com/lucidchart/open/relate/StatementPreparer.scala]: What does this mean?

Comment (pauldraper) [inline src/main/scala/com/lucidchart/open/relate/StatementPreparer.scala]: > It happened

Ick.

**Comment by matthew-lucidchart** [comment]
Is there any way to test these changes? integration tests? unit tests?

What does *gracefully* mean in practice? Does the query stop on the database server?

**Comment by pauldraper** [comment]
It depends on the JDBC implementation at to whether it just stops waiting, or it actually halts the server-side query.

In the case of MySQL, it stops the DB query if you are using 5.0+:

> Connector/J 5.0.0 and later include support for both Statement.cancel() and Statement.setQueryTimeout(). Both require MySQL 5.0.0 or newer server, and require a separate connection to issue the KILL QUERY statement. In the case of setQueryTimeout(), the implementation creates an additional thread to handle the timeout functionality.

**Comment by gregghz** [comment]
I've tested them manually ... not sure how we would unit test it, but an integration test with an active database should be possible with a bit of work.

I tested: `sql"SELECT sleep(1000000)".withTimeout(1).execite()` and after the timeout the query was not still running on the server.

**Comment by pauldraper** [comment]
@gregghz, I suppose you could run that query and then run `show processlist`, and see if there is only one record.

We have integration tests. You have to run them specially outside of SBT. @msiebert knows about that.

**Comment by pauldraper** [comment]
@gregghz, or just write a simple unit test that checks the value of `getTimeout()`. As long as that happens, relate has done its job.

A bit trivial, but then again, so is relate :)

**Comment by gregghz** [comment]
the unit test is tricky because relate doesn't expose the PreparedStatement in the API ... this makes it difficult to check the state of the statement in a unit test.

**Comment by gregghz** [comment]
https://github.com/lucidsoftware/relate-benchmarks

@pauldraper Is that the integration tests you're talking about, or are you thinking of something else?

**Comment by pauldraper** [comment]
> the unit test is tricky because relate doesn't expose the PreparedStatement in the API ... this makes it difficult to check the state of the statement in a unit test.

There are a couple approaches:
* make the PreparedStatement visible by virtue of being in the same package, or being a subclass of a relate class
* use a mock PreparedStatement class, and ensure the correct method is called.

The second is the recommended approach. Mockito makes it less verbose.

The integration test can be run from [it-test.sh](https://github.com/lucidsoftware/relate/blob/master/it-test.sh).

**Pull request merged**